Introduce read-only ContentProvider for cards

joserebelo opened the pull request:
Implement a simple ContentProvider that provides read-only access to the cards. This allows for external apps to read the cards and groups.

Read access is protected by the dangerous permission `me.hackerchick.catima.READ_CARDS`

Gadgetbridge PR with the matching integration (still very much a WIP for experimenting): https://codeberg.org/Freeyourgadget/Gadgetbridge/pulls/2953

Marking as draft since:
- Only for feedback / visibility for now
- Needs some docs
- Need to check if and how to create some tests

joserebelo commented:
It's obviously a problem that we're exposing the internal DB structure, so breaking changes on it break the ContentProvider (#1046 comes immediately to mind).

I'll probably introduce some translation layer to prevent that.

TheLastProject commented:
This is quite cool! Feels much cleaner than my intent idea, thanks for picking this up!

The insert, update and delete methods returning 0 seems quite weird to me. Is this part of the default ContentProvider contract, to return the amount of entries edited and therefore you're always just returning 0 because only reading is allowed? I think having a comment there explaining it would be help (how would making it writable work? How would it check the permission?)

> Alternatively, I can wait for https://github.com/CatimaLoyalty/Android/pull/1046 to get merged and assume the DB structure will remain stable after that (new columns are non-breaking)

I think waiting for #1046 first makes sense. Aside from that it should be fairly safe to assume things will be non-breaking as the database rarely changes.

However, just to be future proof, we may want to consider adding some kind of API level somewhere and put some thin layer between the database and the output so that if we ever have a breaking database change that could be breaking we could have a translation layer ensure the API won't change.

joserebelo commented:
> Is this part of the default ContentProvider contract, to return the amount of entries edited and therefore you're always just returning 0 because only reading is allowed?

Yes, it's recommended in the [documentation](https://developer.android.com/guide/topics/providers/content-provider-creating.html):
> Although you must implement these methods, your code does not have to do anything except return the expected data type. For example, you may want to prevent other applications from inserting data into some tables. To do this, you can ignore the call to insert() and return 0.

I'll add a comment mentioning this.

> how would making it writable work? How would it check the permission?

There's an extra `android:writePermission`. I haven't tried it, but I'm assuming it gets checked automatically by Android when accessing the write methods, just as the readPermission did for the query. Now that you mention it, since I haven't declared it, I'm not sure if any app would be able to call the write methods as-is, even though they don't do anything :thinking: 

Maybe it wouldn't be a bad idea to add the permission right now anyway.

> However, just to be future proof, we may want to consider adding some kind of API level somewhere and put some thin layer between the database and the output so that if we ever have a breaking database change that could be breaking we could have a translation layer ensure the API won't change.

Yup, it's always better to have some sort of API versioning before you actually need it. I'll add a simple `/version` uri with a single record containing a major + minor.

joserebelo commented:
@TheLastProject what are your thoughts regarding merging this before the group IDs change?

I can just add the version uri, assume that the breaking change will happen eventually, and add that to the content provider documentation.

TheLastProject commented:
I think that sounds fine, for now the only users will likely be Gadgetbridge and those users can probably deal with that happening in the future. Just needs some kind of version URI yeah, so integrations can deal with this.

joserebelo commented:
Updated:
- Added docs
- Added a thin layer to hide internal columns (zoom level does not seem relevant)
- Added some basic unit tests to detect breaking changes

Should be ready for a first review.

TheLastProject commented:
Just for my understanding for when I finally find the time to review this (life has been absurdly busy the past month), is https://codeberg.org/Freeyourgadget/Gadgetbridge/pulls/2953 still WIP because the Catima side hasn't been merged yet or because you still expect more changes to be necessary?

joserebelo commented:
@TheLastProject no rush. The pull request on GB is marked as wip mostly because this hasn't been merged yet, just in case we end up doing any breaking changes here.

TheLastProject reviewed (changes requested):
Reading the code seems clear enough and none of it touches the rest of Catima so I'm not really worried about my ability to maintain it (especially because Gadgetbridge will likely be the main user).

I do think it might be good to add an option in settings for "Allow other apps to access my data" so those paranoid can disable this feature completely.

How does requesting for this go? I suppose this is a runtime permission, not an install-time permission?

joserebelo commented on app/src/main/AndroidManifest.xml:
```diff
@@ -2,6 +2,13 @@
+        android:description="@string/permissionReadCardsDescription"
+        android:icon="@drawable/ic_launcher_foreground"
+        android:label="@string/permissionReadCardsLabel"
+        android:name="me.hackerchick.catima.READ_CARDS"
```
> but IIRC this could cause installation conflicts from the regular and debug builds, not allowing both at the same time?

That only happens if there's a conflict between content provider authorities. In the AndroidManifest I set the authority to `${applicationId}.contentprovider.cards`, so both channels can be installed in parallel.

I did forget about the permission. Just tested it, and granting it once on either channel grants it on both apps. Do you think this is OK?

TheLastProject replied:
Sounds fine to me, yeah

joserebelo commented:
> Reading the code seems clear enough and none of it touches the rest of Catima so I'm not really worried about my ability to maintain it (especially because Gadgetbridge will likely be the main user).

I would also be up for fixing issues that come up - just reach out if I do miss any issue :)

> I do think it might be good to add an option in settings for "Allow other apps to access my data" so those paranoid can disable this feature completely.

Should this be disabled by default? If enabled by default (or even if not), I'll probably also include a summary with `Apps will still have to request permissions to access the data`, otherwise it sounds like access is just free for all. 

> How does requesting for this go? I suppose this is a runtime permission, not an install-time permission?

That is correct - this is a runtime permission.

TheLastProject commented:
> I would also be up for fixing issues that come up - just reach out if I do miss any issue :)

Sounds good :)

> Should this be disabled by default? If enabled by default (or even if not), I'll probably also include a summary with Apps will still have to request permissions to access the data, otherwise it sounds like access is just free for all.
> 
> That is correct - this is a runtime permission.

Given it is a runtime permission and people have to explicitly allow it (and it's not just silently allowed when installing a new app) I think enabling it by default is fine. If there is any reason to believe people are being tricked into allowing it from malicious apps we can always start disallowing it by default. Adding your note sounds good.

joserebelo commented:
@TheLastProject small problem - runtime permissions were only included in Android 6.0 (api level 23) and above. Catima's minSdkVersion is 21, just like Gadgetbridge, so the permission is granted silently on Android 5.0.

TheLastProject commented:
Hmm, that's a tough one. Looking at Google Play there are currently 121 Catima users on 5.0 and 5.1, about 0.3% of users. Of course, I have no statistics for F-Droid, but even F-Droid itself dropped Android 5 support recently.

So, given the following:
1. I believe data privacy is more important than new features
2. The amount of Android 5 users is dropping constantly and at ~0.3% currently (Google Play statistics)
3. We are going to have to drop Android 5 support at some point (we're already dealing with some Catima dependencies dropping support like #1204)

I think it makes sense to do the minimal amount of work to keep Android 5 user's data safe and enable the new feature for everyone else. So, just not having the feature on Android 5 sounds like the best way to go. Given Catima is open source, if some Android 5 user really really really wants this feature and understands the risks they could rebuild it with the feature enabled.

joserebelo commented:
Disabled the content provider and hid the preference on SDK < 23.

TheLastProject commented on app/src/main/res/values/strings.xml:
```diff
@@ -115,7 +117,10 @@
     <string name="settings_key_keep_screen_on" translatable="false">pref_keep_screen_on</string>
     <string name="settings_disable_lockscreen_while_viewing_card">Prevent screen lock</string>
+    <string name="settings_allow_content_provider_read_title">Allow other apps to access my data</string>
+    <string name="settings_allow_content_provider_read_summary">Apps will still have to request permissions to be granted access</string>
```
```suggestion
    <string name="settings_allow_content_provider_read_summary">Apps will still have to request permission to be granted access</string>
```

TheLastProject commented:
Thanks!

obfusk commented:
This looks nice! I do wonder what will happen when/if we finally implement proper group IDs?

I guess we could keep the current API and simply hide the IDs from the content provider. Is it possible to implement both that and a newer version of the API that does include it or can we only provide one API version at a time? Or is that just not worth it compared to the relatively trivial change to just use the updated API?

Just wondering if this is something to keep in mind for anyone using the content provider or something to worry about later :)

joserebelo commented:
@obfusk when that happens, I was thinking of just going with the breaking change and bumping the API to 2.0.

> I guess we could keep the current API and simply hide the IDs from the content provider. 

True, but the semantics might change slightly. Right now, the `_id` of groups is actually the group name, and later it will map to an actual ID.

While we could definitely maintain 2 versions of the content provider, I don't think it's worth the extra effort.

Future breaking changes can probably be prevented by manipulating the data on the content provider, but these group ID changes are trickier.

TheLastProject commented:
Given right now Gadgetbridge is the only consumer of this and Gadgetbridge users tend to be fairly tech-savvy I think breaking this is okay especially because we can coordinate releases somewhat. Long term we probably do want to look at backwards compatibility (at least temporarily). But waiting for perfection already delayed this feature for almost a year so it seemed better to get something useful out of the door :)

By the way, please ping me when screenshots or news or so about this feature gets published anywhere, I'd love to see how this works (not having Zepp OS myself, just Bangle.js) :)

joserebelo commented:
@TheLastProject some screenshots below. It's nothing fancy, just a simple preferences screen, but I think it serves as pretty usable MVP.

I was thinking of proposing some intent broadcast from Catima when cards are added / updated, so that Gadgetbridge can get notified of changes and sync the cards accordingly, but haven't thought too much into that yet.

<details>

<summary>Screenshots</summary>

Catima installed

![catima_installed](https://github.com/CatimaLoyalty/Android/assets/1586795/4b085141-dede-4dff-a6d7-5070d1eb9ca8)

Catima installed, but not compatible (no content provider / incompatible version)

![catima_not_compatible](https://github.com/CatimaLoyalty/Android/assets/1586795/7fc1bf0b-221a-4d47-a79f-cb5579da37ba)

Catima not installed

![catima_not_installed](https://github.com/CatimaLoyalty/Android/assets/1586795/a140daf0-d0d6-419c-a6df-a80d0fb5beb8)
</details>
